### ISSUE TITLE
[dynamo]: `assert counter.frame_count == 1` is a bad practice, prefer `counters['graph_breaks']` or `fullgraph=True`

### ISSUE THREAD

**Issue by jon-chuang** [pytorch/pytorch#110730]
### 🐛 Describe the bug

First identified: https://github.com/pytorch/pytorch/pull/110683/files?diff=split&w=0#r1349139286

Bad uncaught graph breaks:

Graph break happens at beginning of function.
```python

from torch._dynamo.testing import CompileCounter
from torch._dynamo.utils import counters

counters.clear()
counter = CompileCounter()

def f(x):
    torch._dynamo.graph_break()
    x += 1
    return x

compiled_f = torch._dynamo.optimize(counter)(f)

x = torch.tensor([1])
compiled_f(x)

print("frame count", counter.frame_count)
print("Graph breaks", len(counters["graph_break"]))
assert counter.frame_count - 1 == len(counters["graph_break"])
```

Graph break occurs at end of function:
```python
from torch._dynamo.testing import CompileCounter
from torch._dynamo.utils import counters

counters.clear()
counter = CompileCounter()

def f(x):
    x += 1
    return torch._dynamo.graph_break()

compiled_f = torch._dynamo.optimize(counter)(f)

x = torch.tensor([1])
compiled_f(x)

print("frame count", counter.frame_count)
print("Graph breaks", len(counters["graph_break"]))
assert counter.frame_count - 1 == len(counters["graph_break"])
```

CC: @voznesenskym @ezyang 

### Versions

main

**Comment by ezyang** [comment]
@jon-chuang if you want to send a PR that adds a lint, or even just comments all the pre-existing bad usages with a "don't copy me, see issue #110730" that would be great, would help stem the bleeding

**Comment by ezyang** [comment]
I think, if your desire is to assert that there are no graph breaks, fullgraph=True is better to express it.

**Comment by voznesenskym** [comment]
asserting frame_count == 1 is perfectly fine - I disagree with the assertion that is is a bad practice. It is, however, incomplete. You need fullgraph=True also. 

You could also imagine a case where fullgraph=True but frame_count is 0 - and everything passes, but the backend compiler was never actually passed a graph.

**Comment by voznesenskym** [comment]
I think what we probably want is an `exactly_k_graphs` compiler backend or util that abstracts away all these notions into a single thing we know works soundly and is easy to use in tests.

**Comment by mlazos** [comment]
I don't think these cases are bugs at all tbh, all of these graph breaks can be safely ignored?

**Comment by jon-chuang** [comment]
> I don't think these cases are bugs at all tbh, all of these graph breaks can be safely ignored.

Yes, but I think the issue is that sometimes it is abused to mean "no graph breaks" which is not true in these corner cases.

**Comment by voznesenskym** [comment]
> Yes, but I think the issue is that sometimes it is abused to mean "no graph breaks" in various tests, which is not true in the above corner cases.

corner cases exist, but by and large its used correctly

> There is no bug in the API, just bugs in the usage.

However, A api that can be misused is a bug in and of itself. Imo, let's make that util, move everyone over to it, and declare success ;)

Thanks for finding the edge cases there, I agree that the whole situation is less than ideal, esp as it is testing code.

**Comment by jon-chuang** [comment]
> but by and large its used correctly

yep, I would expect so.
> I think what we probably want is an exactly_k_graphs

Hmm, are the eager code bits at the beginning and end of the fx module also considered "graphs"?

**Comment by zhxchen17** [comment]
Hi @jon-chuang I'm cleaning up the old issues and wondering if you had a PR to address this? Thanks for proposing the improvement btw!
